### ISSUE TITLE
Cannot draw in global space from a child node

### ISSUE THREAD

**Issue by Zylann** [godotengine/godot#5428]
Godot 2.0.3, Windows 10, 64 bits

I am trying to use `_draw()` to display a sort of particle system in global space, because the ParticleSystem2D is not suited for my case.

I don't want to move my node to the root because I prefer to have it part of my objects, so I am using this workaround to counter all local transforms:

``` gdscript
    var inv = get_global_transform().inverse()
    draw_set_transform(inv.get_origin(), inv.get_rotation(), inv.get_scale())
```

It appears to work, but if I scale the parent node the position is incorrect.
Here is a minimal project showing the issue: 
[GlobalSpace.zip](https://github.com/godotengine/godot/files/333645/GlobalSpace.zip)

Is this a bug? Or am I doing something wrong?

**Comment by reduz** [comment]
this is intended, it's not a bug, and your workaround is valid, it's the
same thing particle system uses.

**Comment by Zylann** [comment]
But why are my drawings mirrored across the Y axis? I expected the rectangle to be drawn above the icon (because I draw at get_global_pos())

**Comment by reduz** [comment]
ah right, because you are using the inverse, just do:

`draw_set_transform( get_global_transform() )
`

**Comment by reduz** [comment]
no wait, nevermind i misunderstood what you wanted to do.

why don't you do: 

draw_set_transform( get_global_transform().affine_inverse() ) ?

**Comment by Zylann** [comment]
```gdscript
	var inv = get_global_transform().affine_inverse()
	draw_set_transform(inv.get_origin(), inv.get_rotation(), inv.get_scale())
```
Same result. What is the difference?

**Comment by reduz** [comment]
ahh, can you try this instead and let me know if it works?

VS.canvas_item_add_set_transform( get_canvas_item(),
get_global_transform().affine_inverse() )

**Comment by Zylann** [comment]
`Invalid call. Nonexistent function 'canvas_item_add_set_transform' in base 'VisualServer'.`
This function is not accessible in GDScript.

**Comment by reduz** [comment]
hm :( I thought GDScript could do this, didn't realize there was not a
function to set a matrix.
Nothing to do then, will add it now to HEAD, but with your current version
I can't think of how to do it.

**Comment by reduz** [comment]
If it works for your case, you could do
set_as_toplevel(true) in the node, and it should draw in global coordinates
(parent transform will not affect it)

**Comment by Zylann** [comment]
I still need the position, orientation and scale of the node, because they are used when I create particles. However, simulation happens in global space.

the VS call works in the minimal project (you mispelled the function, there is no `add` in the name), but in my game it still doesn't looks right. I'm trying to figure out why.

**Comment by reduz** [comment]
you can check particles_2d.cpp  and how it draws them globally

**Comment by Zylann** [comment]
```gdscript
	VS.canvas_item_set_transform( get_canvas_item(), get_global_transform().affine_inverse() )
	
	for p in _particles:
		draw_line(p.pos, p.pos + p.dir*length, color, width)
```
This is how it looks ingame https://youtu.be/hdcOb2pcpwc
Not good :s

**Comment by Zylann** [comment]
particles_2d.cpp uses `VisualServer::get_singleton()->canvas_item_add_set_transform(ci,xform);`,
which is inaccessible from GDScript. I don't see what it does compared to `canvas_item_set_transform`.

**Comment by reduz** [comment]
i commited to github HEAD the function  draw_set_transform_matrix(), are
you compiling godot yourself?

**Comment by Zylann** [comment]
```
d:\projets\info\godot\engine\link_bug\godot\modules\cscript\godot_c.h(9): warning C4067: jetons inattendus aprŠs la directive du pr‚processeur - saut de ligne attendu
d:\projets\info\godot\engine\link_bug\godot\modules\cscript\godot_c.h(210): error C2059: erreur de syntaxeÿ: '__declspec(dllimport)'
d:\projets\info\godot\engine\link_bug\godot\modules\cscript\godot_c.h(235): error C2059: erreur de syntaxeÿ: '__declspec(dllimport)'
d:\projets\info\godot\engine\link_bug\godot\modules\cscript\godot_c.h(256): error C2059: erreur de syntaxeÿ: '__declspec(dllimport)'
d:\projets\info\godot\engine\link_bug\godot\modules\cscript\godot_c.h(294): error C2059: erreur de syntaxeÿ: '__declspec(dllimport)'
d:\projets\info\godot\engine\link_bug\godot\modules\cscript\godot_c.h(316): error C2059: erreur de syntaxeÿ: '__declspec(dllimport)'
d:\projets\info\godot\engine\link_bug\godot\modules\cscript\godot_c.h(338): error C2059: erreur de syntaxeÿ: '__declspec(dllimport)'
d:\projets\info\godot\engine\link_bug\godot\modules\cscript\godot_c.h(424): error C2059: erreur de syntaxeÿ: '__declspec(dllimport)'
d:\projets\info\godot\engine\link_bug\godot\modules\cscript\godot_c.h(444): error C2059: erreur de syntaxeÿ: '__declspec(dllimport)'
d:\projets\info\godot\engine\link_bug\godot\modules\cscript\godot_c.h(449): error C2059: erreur de syntaxeÿ: '__declspec(dllimport)'
d:\projets\info\godot\engine\link_bug\godot\modules\cscript\godot_c.h(454): error C2059: erreur de syntaxeÿ: '__declspec(dllimport)'
d:\projets\info\godot\engine\link_bug\godot\modules\cscript\godot_c.h(526): error C2059: erreur de syntaxeÿ: '__declspec(dllimport)'
d:\projets\info\godot\engine\link_bug\godot\modules\cscript\godot_c.h(558): error C2059: erreur de syntaxeÿ: '__declspec(dllimport)'
d:\projets\info\godot\engine\link_bug\godot\modules\cscript\godot_c.h(559): error C2059: erreur de syntaxeÿ: '__declspec(dllimport)'
scons: building terminated because of errors.
```
woops

**Comment by Zylann** [comment]
Does it builds on your side?

**Comment by neikeq** [comment]
@Zylann Builds fine for me. It may happen only on windows. Consider making a new issue.

**Comment by vnen** [comment]
@Zylann this PR fixes the build problem: #5434

**Comment by Zylann** [comment]
@vnen I created an issue while you answered https://github.com/godotengine/godot/issues/5435

**Comment by Zylann** [comment]
Ok, I can build now, thanks :)

So, this issue has been closed just to add the `_matrix` version of `draw_set_transform`? But it doesn't solves the issue, it just simplifies how the function is called.

Now I can see there are multiple ways of drawing to global space, but I don't understand why are there different?
1)
```gdscript
        # Works, unless the parent is scaled. If the parent has (-1,1) scale, drawings are mirrored across the origin, so it doesn't really counters the transform...
	draw_set_transform_matrix(get_global_transform().affine_inverse())
```
2)
```gdscript
        # Also works, but seems to be originally unrelated to drawing? (also looks harder to understand, I would never have found that code)
	VS.canvas_item_set_transform( get_canvas_item(), get_global_transform().affine_inverse() )
```

**Comment by vnen** [comment]
> also I don't see what "affine" means here

The tutorial about Matrices and Transforms explain this: http://docs.godotengine.org/en/latest/tutorials/matrices_and_transforms.html#affine-inverse

----

Reopening since it seems to not be solved.

**Comment by vnen** [comment]
Well, now I tested and it works fine for me. @Zylann are you sure you're getting wrong results? Can you add screenshots or another project with the tests?

**Comment by vnen** [comment]
I used `draw_set_transform_matrix(get_global_transform().affine_inverse())` and `-2,2` scale for the Sprite. It is drawing in the correct place:

![godot windows tools 64_2016-06-26_20-14-42](https://cloud.githubusercontent.com/assets/5599796/16365578/bdeb6232-3bda-11e6-8d13-72109e9382c0.png)

**Comment by Zylann** [comment]
Indeed, that's strange. Because those two calls give me different results:

```gdscript
    # The drawing is a the wrong place
	var inv = get_global_transform().affine_inverse()
	draw_set_transform(inv.get_origin(), inv.get_rotation(), inv.get_scale())
```
```gdscript
    # The drawing is a the correct place
	draw_set_transform_matrix(get_global_transform().affine_inverse())
```
Aren't they supposed to do the exact same thing? (one being a shorthand)

**Comment by bojidar-bg** [comment]
@Zylann it is possible that the matrix transformation can't be expressed by just pos, rot, and scale, like a rotated object in a non-uniformly scaled parent. I'm not sure that's the whole problem though...

**Comment by Zylann** [comment]
@bojidar-bg good point. My object is not rotated, thought.

**Comment by reduz** [comment]
so, is this still a bug? I also don't want to state the obvious but you could put the node with the shoots at a higher hierarhcy too :P

**Comment by Zylann** [comment]
Yeah I think the new function accepting a matrix solves my problem with the latest version of Godot.
I still don't want to move the node to top level because that's too much constraint (and I need to draw in global space very often, mostly for debugging, tooling and world-space simulations like particles).

**Comment by reduz** [comment]
also remember that you can use set_as_toplevel() and the node will draw in
world-space

**Comment by Zylann** [comment]
Yeah but I don't want all things from the node to be drawn in world space, only some of those that I draw by code.

**Comment by vnen** [comment]
@reduz the problem I see here is that `transform.get_scale()` only return positive values even when the scale is negative. That's why it gets drawn in the wrong place when using ` draw_set_transform(inv.get_origin(), inv.get_rotation(), inv.get_scale())`.

**Comment by DualMatrix** [comment]
This seems to be the same problem as #17405

**Comment by akien-mga** [comment]
Looks like it, closing as part of the discussion in this issue was resolved in 130b83b, the rest (negative scale issue) is indeed covered in #17405.
